# 5. Data Consistency Check
## Table of contents

***

**`1. Import libraries`**<br/>

**`2. Create reusable path`**<br/>

**`3. Import data`**<br/>

**`4. Example of identifying mixed-type columns`**<br/>

**`5. Product dataframe consistency check`**<br/>
- _5.1 Locate missing values (NULL)_<br/>
- _5.2 Isolate missing values_<br/>
- _5.3 Address missing values_ <br/>
- _5.4 Isolate the duplicates_<br/>
- _5.5 Remove the duplicates_<br/>

**`6. Order dataframe data consistency check`**<br/>
- _6.1 Investigate the accuracy of all columns_<br/>
- _6.2 Identifying mixed type column_ <br/>
- _6.3 Check for missing values_ <br/>
- _6.4 Address the missing value_<br/>
- _6.5 Check for the duplicate values_<br/>
- _6.6 Addressing dulplicates_<br/>

**`7. Export final dataframe for both 'product' & 'order' dataframes`**
***

### 1. Import libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 2. Create reusable path

In [4]:
# Create path
path = r'/Users/bonnieandbanana/Desktop/A New Career/CareerFoundry/Data Immersion/A4_Python/ Instacart Basket Analysis Project (IC)'

### 3. Import Data

In [5]:
# Import 'product.csv' dataset from original data
df_prods = pd.read_csv(os.path.join(path, '02 data' , 'original data', 'products.csv'))

# Import 'orders_wrangled.csv' dataset from prepared data
df_ords = pd.read_csv(os.path.join(path, '02 data' , 'prepared data', 'orders_wrangled.csv'), index_col = 0)

### 4. Example of identifying mixed-type columns

In [6]:
# Create a test dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed-type column
df_test ['mix'] = ['a', 'b', 1 , True]

In [8]:
# Check output
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Identifying mixed-type column within the dataframe 'df_test'
for col in df_test.columns.tolist():
    weird = (df_test[[col]] . applymap(type) != df_test [[col]] . iloc[0] . apply(type)) . any(axis = 1)
    if len (df_test [weird]) > 0 :
             print (col)

mix


### 5. Product dataframe consistency check

#### _5.1 Locate missing value (NULL)_

In [10]:
# Step 1: look for missing value
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

#### _5.2 Isolate missing values_

In [11]:
# Step 2: create a subset of the dataframe containing only the NaN values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
# Check the number of rows 
df_prods.shape

(49693, 5)

#### _5.3 Address missing value_

<p>method 1: create new variable to flag the missing value
<p>method 2: impute the value with mean or median
<p>method 3: remove the row contains missing values (demonstrated below)

In [14]:
# Step 3: create a new dataframe that filters the missing values ou in 'product_name' column
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

#### _5.4 Isolate the duplicates_

In [16]:
# Step 1: create a subset that only contains the dulplicates
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [17]:
# Check the number of rows that are duplicated
df_dups.shape

(5, 5)

In [18]:
# Before dropping dulplicates, check the number of row of the current dataframe
df_prods_clean.shape

(49677, 5)

In [19]:
# Check output
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

#### _5.5 Remove the duplicates_

In [20]:
# Drop the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
# Check the row of the dataframe after the removal of the duplicates
df_prods_clean_no_dups.shape

(49672, 5)

### 6. Order dataframe data consistency check

#### _6.1 Investigate the accuracy of all columns_

In [22]:
# Obtain descriptive information
df_ords.describe()

order_id       user_id  order_number  order_day_of_the_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06           3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01           2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01           2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00           0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00           1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01           3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01           5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02           6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

**interpretation:**

- **order_number**<br>
Maximum order number indicates that the order number had reached 100. <br>
- **order_day_of _week**<br>
Minimum value '0' and maximum value '6' show that there're seven numbers indicating the 7 days of a week.<br>
- **order_hour_of_day**<br>
Minimu value '0' and maximum value '23 total the 24 hours of a day.<br>
- **days_since_last_order**<br>
Minimum value '0' indicates that some orders were placed on the same day; maximum value of '30' shows the 30 days of a month.<br>

**Note:** The counts of column 'days_since_last_order is less than the rest of the columns, indicating there're missing values in the column.<br>
<br>
**Data Type:**<br>
The order_id and user_id should be string data type rather than numerical.<br>

#### _6.2 Identifying mixed type column_

In [23]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]] . applymap(type) != df_ords [[col]] . iloc[0] . apply(type)) . any(axis = 1)
    if len (df_ords [weird]) > 0 :
             print (col)

**Obervsation:**<br>
There's no column with mix-type data present in df_ords

In [24]:
# Check the data type of all the columns
df_ords.dtypes

order_id                   int64
user_id                    int64
order_number               int64
order_day_of_the_week      int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

#### _6.3 Check for missing values_

In [25]:
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
order_day_of_the_week         0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

**Obervation:**<br>
There're '206209' rows with missing values in column of days_since_last_order, the rest of the columns do not have missing values which justify the investigation from column accuracy check.

>**Appropriate method evaluation**<br>
>The column of 'days_since_last_order' would not have any values if the entry is from the first orders, base on this reason, the missing values are considered relevent.<br>
>After careful evaluation, to be able to continue the current analysis, the missing values should be separated from the dataframe by creating a subset that contains only the nulls.

#### _6.4 Address the missing value_

In [26]:
# Step 1: separate the NaN value in the column
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [27]:
df_ords_nan

order_id  user_id  order_number  order_day_of_the_week  \
0         2539329        1             1                      2   
11        2168274        2             1                      2   
26        1374495        3             1                      1   
39        3343014        4             1                      6   
45        2717275        5             1                      3   
...           ...      ...           ...                    ...   
3420930    969311   206205             1                      4   
3420934   3189322   206206             1                      3   
3421002   2166133   206207             1                      6   
3421019   2227043   206208             1                      1   
3421069   3154581   206209             1                      3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

**Observation**<br>
A verification of whether all the entries with NaN value in column 'days_since_last_order' are due to '1' order is needed.

In [28]:
# Step 2: filter the dataframe where 'order_number' equals 1
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [29]:
# Step3: check if all of these have missing values in 'days_since_last_order'
df_first_orders_missing = df_first_orders['days_since_last_order'].isna().all()

In [30]:
# Step 4: validate
df_first_orders_missing

True

In [31]:
# Print the complete result
print("All first orders have missing value in 'days_since_last_order':", df_first_orders_missing)

All first orders have missing value in 'days_since_last_order': True


In [32]:
# Step 5: Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_last_order'].isnull()

In [33]:
# Step 6: Check the new inserted column
df_ords

order_id  user_id  order_number  order_day_of_the_week  \
0         2539329        1             1                      2   
1         2398795        1             2                      3   
2          473747        1             3                      3   
3         2254736        1             4                      4   
4          431534        1             5                      4   
...           ...      ...           ...                    ...   
3421078   2266710   206209            10                      5   
3421079   1854736   206209            11                      4   
3421080    626363   206209            12                      1   
3421081   2977660   206209            13                      1   
3421082    272231   206209            14                      6   

         order_hour_of_day  days_since_last_order  first_order  
0                        8                    NaN         True  
1                        7                   15.0        False  
2                       12                   21.0        False  
3                        7                   29.0        False  
4                       15                   28.0        False  
...                    ...                    ...          ...  
3421078                 18                   29.0        False  
3421079                 10                   30.0        False  
3421080                 12                   18.0        False  
3421081                 12                    7.0        False  
3421082                 14                   30.0        False  

[3421083 rows x 7 columns]

In [34]:
# Step 7: rename df_ords to df_ords_clean
df_ords_clean = df_ords

In [35]:
# Step 8: check data-type of new column
df_ords_clean['first_order'].dtype

dtype('bool')

In [36]:
# Step 9: double check if the dataframe contains the right number of rows of NaN value
df_ords_clean.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
order_day_of_the_week         0
order_hour_of_day             0
days_since_last_order    206209
first_order                   0
dtype: int64

**Observation**<br>
The NaN values are addressed appropriately, clearly flagged in 'first_order' column which helps the client to understand why the NaN values existed in 'days_since_last_order' column<br>
<br>
The removal of NaN value in 'days_since_last_order' could result in inaccurate analysis as the information is vital for understanding customer behavior.

#### _6.5 Check for the duplicate values_

In [37]:
# Step 1: create a subset of duplicates from df_ords
df_ords_clean_dup = df_ords_clean[df_ords_clean.duplicated()]

In [38]:
# Step 2: check the duplicates in the subset created
df_ords_clean_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_the_week, order_hour_of_day, days_since_last_order, first_order]
Index: []

In [42]:
# Step 3: rename 'df_ords_clean_dup' to 'df_ords_clean_no_dup'
df_ords_clean_no_dups = df_ords_clean

#### _6.6 Addressing dulplicates_

**Observation**<br>
The dataframe created to check duplicates has returned empty, meaning there are no duplicates in the data set.

### 7. Export final dataframe for both 'product' & 'order' dataframes

In [43]:
df_ords_clean_no_dups.to_csv(os.path.join(path, '02 data','prepared data','orders_cleaned.csv'))
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 data', 'prepared data', 'products_cleaned.csv'))